added a tutorial of car paint shop problem

bluest93 opened the pull request:
## Change
Added a tutorial of multi-car paint shop problem 

## Description 
1. Added a tutorial notebook of "QAOA for Two-Color Multi Car Paint Shop Optimization"
2. Revised the `index.md` 
3. Revised the `_toc.yml`

yuyamashiro reviewed (changes requested):
- Please include a brief explanation of the model.
- It appears that openjij is not being used. Please remove any unnecessary imports.
- The model's Element `e` also appears to be unnecessary. Please remove it if you do not need it.
- Please add simple visualization code for the optimization results so that we can see what the results are.

bluest93 commented:
@yuyamashiro I updated the notebook regarding to the four suggestions above. Also, I changed the problem to minimization, since `sampleset.lowest()` is used in the result part to show the optimal solution.

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+    "id": "i9nXiSl26DJK"
+   },
+   "source": [
+    "# Quantum Alternating Operator Ansatz for Two-Color Multi Car Paint Shop Problem\n",
```
Please change this as QAOA because this is not Quantum Alternating Operator Ansatz.

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+   },
+   "source": [
+    "# Quantum Alternating Operator Ansatz for Two-Color Multi Car Paint Shop Problem\n",
+    "The goal of the multi-car paint shop problem is to reduce the number of color switches between cars in a paint shop line during the manufacturing process, which is recognized as an NP-hard problem.\n",
```
You should cite original paper.
https://arxiv.org/abs/2109.07876

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+    "The **cost** of the sequence is the number of times the color switches between consecutive cars, which can be formulated as:\n",
+    "\n",
+    "$$\n",
+    "\\text{min} \\quad -\\sum_{i=0}^{N-2} (x_i - 0.5) \\cdot (x_{i+1} - 0.5)\n",
```
This formulation has huge jump because the spin variables $s_i \in \{-1,1\}$ are used in the original formulation.
However, in jijmodeling, we can only use binary varaibles $x_i \in \{0,1\}$, so we use this weird formulation, so I think it it better to start the explanation from spin variables, then change the spin variables into binary variables.

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The order of car intake:  [1, 1, 0, 2, 3, 0, 3, 2]\n",
```
This way to show the result is hard to understand the meaning, so please make more pretty plot. such as using emoji like 🚗 somothing

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+    "    energy = round(sample.eval.objective, ndigits = 3)  \n",
+    "    frequencies[energy] += sample.num_occurrences\n",
+    "\n",
+    "plt.bar(frequencies.keys(), frequencies.values(), width=0.05)\n",
```
I think the size of the bar is narrow.

Jacomichi commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,541 @@
+   "metadata": {},
+   "source": [
+    "## Evaluation using classical algorithms\n",
+    "Comparing the best costs by searching through all feasible solutions"
```
> searching through all feasible solutions

We can this as brute force search

Jacomichi reviewed (changes requested):
I have left the comments

bluest93 commented:
@Jacomichi I have modified the notebook upon the requests above, please take a look

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "## Expressing the problem using a mathematical model\n",
+    "\n",
+    "Given a set of cars $ X = \\{x_1, x_2, \\dots, x_n\\} $ that need to be painted in one of two colors, denoted as $ x_i=0 $ and $ x_i=1 $,  for each $ i = 1, \\dots n$ . The objective is to minimize the number of color switches (i.e., changes from $ 0 $ to $ 1 $ or from $ 1 $ to $ 0 $) between consecutive cars in the sequence.\n",
```
```suggestion
    "Given a set of cars $ X = \\{x_1, x_2, \\dots, x_n\\} $ that need to be painted in one of two colors, denoted as $ x_i=0 $ and $ x_i=1 $,  for each $ i = 1, \\dots n$. The objective is to minimize the number of color switches (i.e., changes from $ 0 $ to $ 1 $ or from $ 1 $ to $ 0 $) between consecutive cars in the sequence.\n",
```

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "\\text{min} \\quad -\\sum_{i=0}^{N-2} (x_i - 0.5) \\cdot (x_{i+1} - 0.5)\n",
+    "$$\n",
+    "\n",
+    "In the Two-Color Multi-Car Paint Shop Problem, the goal is to minimize the number of color switches while meeting specific constraints. In this notebook, we consider the constraint that the scheduled number of cars in each of the two colors per model (total number of models is $M$) must be met.\n",
```
```suggestion
    "In the Two-Color Multi-Car Paint Shop Problem, the goal is to minimize the number of color switches while meeting specific constraints. In this notebook, we consider the constraint that the scheduled number of cars in each of the two colors per model (the total number of models is $M$) must be met.\n",
```

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+    }
+   ],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
```
I change the fontsize 
```
plt.title("Change of Cost", fontsize=15)
plt.xlabel("Iteration", fontsize=15)
plt.ylabel("Cost", fontsize=15)
plt.xscale("log")
plt.xlim(1, result.nfev)
plt.plot(cost_history, label="Cost", color="#2696EB")
plt.show()
```

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+    }
+   ],
+   "source": [
+    "\n",
```
I think no need the space here

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "    energy = round(sample.eval.objective, ndigits = 3)  \n",
+    "    frequencies[energy] += sample.num_occurrences\n",
+    "\n",
+    "plt.bar(frequencies.keys(), frequencies.values(), width=0.5)\n",
```
Also chcnge the ticker for X.
```
plt.bar(frequencies.keys(), frequencies.values(), width=0.25)
plt.xticks(sorted(np.array(list(frequencies.keys()))))
plt.xlabel('Objective', fontsize=15)
plt.ylabel('Frequency', fontsize=15)
plt.show()
```

ahao27 commented on docs/tutorial/multi_car_paint.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "metadata": {},
+   "source": [
+    "## Evaluation using classical algorithms\n",
+    "Comparing the best costs by brute force search"
```
```suggestion
    "Comparing the best costs by brute force search."
```

ahao27 reviewed (commented):
I have reviewed them, please check it.

bluest93 commented:
@ahao27 Thank you so much. I have modified the notebook based on the requests

ahao27 reviewed (commented):
Last two small things

ahao27 commented:
LGTM

yuyamashiro reviewed (approved):
LGTM

Jacomichi reviewed (approved):
LGTM
